hp_search_benchmarks/nas_search_spaces/darts/cnn/model_def.py

```diff
@@ -0,0 +1,180 @@
+            for node in range(4):
+                for edge in [1, 2]:
+                    edge_ind = self.hparams[
+                        "{}_node{}_edge{}".format(cell, node + 1, edge)
+                    ]
+                    edge_op = self.hparams[
+                        "{}_node{}_edge{}_op".format(cell, node + 1, edge)
+                    ]
+                    cell_config[cell].append((edge_op, edge_ind))
+        print(cell_config)
+        return Genotype(
+            normal=cell_config["normal"],
+            normal_concat=range(2, 6),
+            reduce=cell_config["reduce"],
+            reduce_concat=range(2, 6),
+        )
+
+    def build_model(self) -> nn.Module:
+        genotype = self.get_genotype_from_hps()
+
+        model = Network(
+            self.hparams["init_channels"],
+            10,  # num_classes
+            self.hparams["layers"],
+            self.hparams["auxiliary"],
+            genotype,
+        )
+        print("param size = {} MB".format(utils.count_parameters_in_MB(model)))
+        size = 0
+        for p in model.parameters():
+            size += p.nelement()
+        print("param count: {}".format(size))
+
+        # If loading backbone weights, do not call reset_parameters() or
+        # call before loading the backbone weights.
+        reset_parameters(model)
+        return model
+
+    def optimizer(self, model: nn.Module) -> torch.optim.Optimizer:  # type: ignore
+        self._optimizer = torch.optim.SGD(
+            model.parameters(),
+            lr=self.context.get_hparam("learning_rate"),
+            momentum=self.context.get_hparam("momentum"),
+            weight_decay=self.context.get_hparam("weight_decay"),
+        )
+        return self._optimizer
+
+    def create_lr_scheduler(self, optimizer):
+        self.scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            optimizer, self.context.get_hparam("train_epochs")
+        )
+        step_mode = LRScheduler.StepMode.STEP_EVERY_EPOCH
+        return LRScheduler(self.scheduler, step_mode=step_mode)
+
+    def train_batch(
+        self, batch: Any, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        input, target = batch
+        model.drop_path_prob = (
+            self.hparams["drop_path_prob"] * epoch_idx / self.hparams["train_epochs"]
+        )
+        if batch_idx == 0 or epoch_idx > self._last_epoch:
+            print("epoch {} lr: {}".format(epoch_idx, self.scheduler.get_last_lr()[0]))
+            print("drop_path_prob: {}".format(model.drop_path_prob))
+
+        logits, logits_aux = model(input)
+        loss = self.criterion(logits, target)
+        if self.context.get_hparam("auxiliary"):
+            loss_aux = self.criterion(logits_aux, target)
+            loss += self.context.get_hparam("auxiliary_weight") * loss_aux
+        top1, top5 = utils.accuracy(logits, target, topk=(1, 5))
+
+        self._last_epoch = epoch_idx
```

Comment:
The state of this attribute is not going to be correctly saved across checkpointing / restoring, e.g. if you pause and restore a trial. You could use the new PyTorch callback interface that has the ability `save_state_dict` and `load_state_dict` to achieve this though.

Reply:
I only use it for logging so it's not critical that it's saved correctly